zerver/lib/converters.py

```diff
@@ -162,6 +242,40 @@ def configure_converter(converter: BaseConverter) -> None:
 for target_type in (str, int, float, bool):
     json_converter.register_structure_hook(target_type, structure_isinstance)
 
+
+def structure_union(value: object, target_type: Type[T]) -> T:
+    """A generic structure function for Union[T1, T2, ...]. It works by
+    attempting to structure the value with each type argument of the Union until
+    it succceeds, and raises a UnionError when none of them works.
+    """
+    # Structuring Union[...] is supported by json_converter only, because we
+    # rely on the parsed JSON output to disambiguate among the accepted types.
+    type_args = get_args(target_type)
+    errors = []
+    for type_arg in type_args:
+        try:
+            return json_converter.structure(value, type_arg)
+        except Exception as exc:
+            attach_error_note(exc, f"Structuring as {type_arg} from {target_type}", type_arg)
+            errors.append(exc)
+    raise UnionError(f"{value} cannot be structure as {target_type}", errors, target_type)
+
+
+# To be explicit about what unions are allowed to be handled by the generic
+# structure_union hook, we whitelist them here.
```

Comment:
"we whitelist them here" -> "we specify the exact concrete union types that are necessary in Zulip. (Zulip typically has union types related to backward compatibility of APIs that originally took strings but now take ids.  We generally want to phase out union types where possible.)"

(Some folks object to the terms "whitelist" and "blacklist".  While I don't necessarily agree with their objections, you can usually use more precise language like allow-list and deny-list, or, as in this case, just use a more detailed explanation.)

Reply:
Thanks. I moved `allowed_unions` to the top of the file and added the explanation there.